### PR TITLE
Bug 1370824 - Add Pinning of TopSites to the Activity Stream Panel

### PR DESCRIPTION
**Pull request by farhanpatel**
Opting for a simple approach to pinning topsites. You can pin a site from the Context menu and it'll occupy the first position in the TopSites section.
You cant rearrange topsites. But you can re-pin a site in order to bring it back to the front. 

You'll also be able to pin a site from the menu and other home panels. But that'll be a separate PR. 

This is my first database migration. I opted for using a separate table because the topsite cache table caches favicon metadata along with site data. If we allowed pin sites to exist in the topsite cache we would have to update cached favicon metadata manually for sites.  Imagine if a site is pinned before we load a favicon. if we got favicon later we'd have to manually update the topsite_cache_table.
It only complicated the logic of adding/remove topsites. I don't really understand the downsides of adding a table so this could be worth it. Im fine doing it either way.

Probably my last PR you'll review 😭

### CODE REVIEW

**Comment by farhanpatel** [comment]
### SwiftLint found issues

#### Warnings

File | Line | Reason |
| --- | ----- | ----- |
ActivityStreamPanel.swift | 235 | Force casts should be avoided. 
ActivityStreamPanel.swift | 265 | Force casts should be avoided. 
ActivityStreamPanel.swift | 337 | Force casts should be avoided. 
ActivityStreamPanel.swift | 346 | Force casts should be avoided. 
ActivityStreamPanel.swift | 352 | Force casts should be avoided. 
ActivityStreamPanel.swift | 505 | Force casts should be avoided. 
BrowserTable.swift | 702 | Force casts should be avoided. 
SQLiteHistory.swift | 829 | Force casts should be avoided. 
SQLiteHistory.swift | 872 | Force casts should be avoided. 
SQLiteHistory.swift | 877 | Force casts should be avoided. 
SQLiteHistory.swift | 975 | Force casts should be avoided. 
SQLiteHistory.swift | 995 | Force casts should be avoided. 
SQLiteHistory.swift | 996 | Force casts should be avoided. 
SQLiteHistory.swift | 997 | Force casts should be avoided. 
SQLiteHistory.swift | 998 | Force casts should be avoided. 
SQLiteHistory.swift | 1038 | Force casts should be avoided. 
SQLiteHistory.swift | 1040 | Force casts should be avoided. 
SQLiteHistory.swift | 149 | Limit vertical whitespace to a single empty line. Currently 2. 
TestSQLiteHistory.swift | 1588 | Force casts should be avoided. 
TestSQLiteHistory.swift | 1588 | Force casts should be avoided. 
TestSQLiteHistory.swift | 1456 | Limit vertical whitespace to a single empty line. Currently 2. 
TestSQLiteHistory.swift | 974 | Prefer `_ = foo()` over `let _ = foo()` when discarding a result from a function. 


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: Storage/SQL/SQLiteHistory.swift [hunk 1]
```diff
@@ -136,6 +136,25 @@ extension SQLiteHistory: BrowserHistory {
...
+        return db.runQuery(sql, args: [], factory: SQLiteHistory.iconHistoryMetadataColumnFactory)
+    }
+
+
+    public func addPinnedTopSite(_ site: Site) -> Success { // needs test
+        let now = Date.now()
+        guard let siteID = site.id else {
+            return deferMaybe(DatabaseError(description: "Invalid site \(site.url)"))
+        }
+        return db.run([("INSERT OR IGNORE INTO \(TablePinnedTopSites)(url, pinDate, title, historyID, guid) VALUES (?,?,?,?,?)", [site.url, now, site.title, siteID, site.guid])])
```
Comment (sleroux) [inline Storage/SQL/SQLiteHistory.swift]: Nit: `BrowserDB` has a handy method named `varlist` that generates the (?,?,....) for you

File: StorageTests/TestSQLiteHistory.swift [hunk 2]
```diff
@@ -1,4 +1,4 @@
-/* This Source Code Form is subject to the terms of the Mozilla Public
+ /* This Source Code Form is subject to the terms of the Mozilla Public
```
Comment (sleroux) [inline StorageTests/TestSQLiteHistory.swift]: Nit: Whitespace

File: Client/Frontend/Home/ActivityStreamPanel.swift [hunk 3]
```diff
@@ -624,18 +633,22 @@ extension ActivityStreamPanel: DataObserverDelegate {
...
-            self.hideURLFromTopSites(site.tileURL)
+            self.hideURLFromTopSites(site)
         })
 
         let dismissHighlightAction = ActionOverlayTableViewAction(title: Strings.RemoveFromASContextMenuTitle, iconString: "action_close", handler: { action in
             self.telemetry.reportEvent(.Dismiss, source: pingSource, position: index)
             self.hideFromHighlights(site)
         })
 
+        let pinTopSite = ActionOverlayTableViewAction(title: "Pin TopSite", iconString: "action_share", handler: { action in
```
Comment (sleroux) [inline Client/Frontend/Home/ActivityStreamPanel.swift]: Make sure to move these string into `Strings.swift` for localization.

Comment (sleroux) [inline Client/Frontend/Home/ActivityStreamPanel.swift]: Also it probably makes sense to just be 'Pin Site'

Comment (sleroux) [inline Client/Frontend/Home/ActivityStreamPanel.swift]: Oh and the title should change depending on if the site is pinned or not. If its already been pinned it should say "Unpin site" or something.

Comment (farhanpatel) [inline Client/Frontend/Home/ActivityStreamPanel.swift]: You can still pin a site again if its pinned. Thats how you can move a pinned site to the front of the list if you want to change the order. 
If you want to remove a pinned site its the same as just removing a regular topsite.

**Pull request merged**